src/ophys_etl/utils/array_utils.py

```diff
@@ -85,3 +85,34 @@ def normalize_array(
     normalized -= lower_cutoff
     normalized = np.uint8(normalized * 255 / (upper_cutoff - lower_cutoff))
     return normalized
+
+
+def pairwise_distances(points: np.ndarray) -> np.ndarray:
+    """
+    Calculate all of the pairwise distances between the rows
+    in a np.ndarray
+
+    Parameters
+    ----------
+    points: np.ndarray
+        Shape is (n_points, n_dimensions)
+
+    Returns
+    -------
+    distances: np.ndarray
+        A (n_points, n_points) array. The i,jth element
+        is the Euclidean distance between the ith and jth
+        rows of the input points.
+
+    Notes
+    -----
+    As n_points, n_dimensions approach a few thousand, this is
+    several orders of magnitude faster than scipy.distances.cdist
```

Comment:
in cdist, did you try metric = 'sqeuclidean`? I recall that being significantly faster than `euclidean`

Reply:
I don't see any speed up going to `sqeuclidean`

Reply:
another possibility is `pdist` for this use case.
I'm skeptical that your loop can exceed some compiled C here

Reply:
i recall now, this is probably impacted a lot by what BLAS is used. so, we could be comparing apples and oranges. my experience is that with BLAS provided by MKL, there was no beating cdist

Reply:
```
np result shape (1000, 1000)
pdist shape (499500,)
duration pdist 0.11454200744628906 seconds
duration np 0.025458812713623047 seconds
```
Note: I haven't done the work to untangle the unnecessarily clever shape of `pdist`'s output to verify that I'm getting the same result. But, running on 1000 250 dimensional points, the homebrew numpy implementation is faster.

I have had some bad experiences with scipy. It's just never faster than doing it yourself with numpy.

Reply:
(this is all on my laptop)

Reply:
In the scipy [code](https://github.com/scipy/scipy/blob/v1.7.0/scipy/spatial/distance.py#L2606), they iterate over each row of the 2 inputs, which would make their algorithm slower than this one

Reply:
I created a new conda environment and did my best to make absolutely sure I got the MKL-enabled scipy and numpy. I still see that the numpy implementation is ~ 1 order of magnitude faster. Specifically

```
duration cdist 0.9538030624389648 seconds
duration pdist 0.4869711399078369 seconds
duration np 0.026255130767822266 seconds
```

Here is my `scipy.show_config()`
```
>>> import scipy
>>> scipy.show_config()
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/include']
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/scott.daniel/miniconda3/envs/mkl_test/include']
```
I *think* this means I am using MKL, but the answers I get to "how do I tell if scipy is using MKL" all seem very cagey.

If you want, I can send you my benchmarking script to run on your own system to see if you get a different result.

One thing to consider: because this method is run on individual ROI growth, which is already farmed out to a child process, we need to use the fastest single-thread implementation of this algorithm, since we don't want multiple child processes competing for all of the machine's cores. I am saying this in case the boost in speed due to MKL is related to efficient silent multithreading in scipy and numpy.

Reply:
I see a cross-over in performance where your method wins when the feature vector dimension gets larger than 100-200. Below that, scipy wins.

Reply:
anyway, no change requested